### ISSUE TITLE
PG::Connection.transaction changes interrupt behavior

### ISSUE THREAD

**Issue by thatsmydoing** [ged/ruby-pg#390]
When doing a plain query like

```
conn.async_exec('select pg_sleep(500)')
```
Ctrl-C interrupts the query immediately and exits.

when wrapped in a transaction like

```
conn.transaction do |c|
  c.async_exec('select pg_sleep(500)')
end
```
it will not process any signals until the sleep finishes.

Doing the transaction in pure ruby like so
```
begin
  conn.async_exec('begin')
  conn.async_exec('select pg_sleep(500)')
  conn.async_exec('commit')
rescue SignalException
  conn.cancel
  conn.async_exec('rollback')
end
```
is cancellable and properly rolls back the transaction.

We run into this problem in rails migrations which does use `PG::Connection.transaction` and we can't easily swap in the ruby version.

**Comment by thatsmydoing** [comment]
This patch seems to fix it?
```
diff --git a/ext/pg_connection.c b/ext/pg_connection.c
index 5519798..a012d94 100644
--- a/ext/pg_connection.c
+++ b/ext/pg_connection.c
@@ -2932,6 +2932,7 @@ pgconn_transaction(VALUE self)
 		}
 		else {
 			/* exception occurred, ROLLBACK and re-raise */
+			pgconn_cancel(self);
 			result = gvl_PQexec(conn, "ROLLBACK");
 			rb_pgresult = pg_new_result(result, self);
 			pg_result_check(rb_pgresult);
```
not sure if there's a more proper way of handling it. I suppose you only want to cancel on `SignalException` or if you know that there's still a pending query.

**Comment by larskanis** [comment]
I moved the method to ruby code, so that the signal handling works. However the query is not actively cancelled - this is something the calling code has to do.

You might use `conn.exec` instead of `conn.async_exec` - the latter one is an alias.

**Comment by larskanis** [comment]
I noticed that the transaction still can't be stopped by Ctrl-C. This is because `rescue Exception` catches `Interrupt` and the connection is still in processing state, so that `exec "ROLLBACK"` blocks next.

Adding `block` before `exec` fixes the blocking, so that the application stops when pressing `Ctrl-C` twice:

```
--- a/lib/pg/connection.rb
+++ b/lib/pg/connection.rb
@@ -225,6 +225,7 @@ class PG::Connection
                exec "BEGIN"
                res = yield(self)
        rescue Exception
+               block
                exec "ROLLBACK"
                raise
        else
```

Not sure if that's the best way.

**Comment by thatsmydoing** [comment]
> However the query is not actively cancelled - this is something the calling code has to do.

While I understand that it seems wrong to automatically cancel a query, the next statement is a `ROLLBACK` anyway, why not just cancel it? The behavior would be different for inside vs outside a transaction since interrupting a non-transaction statement does not cancel it but I think that at least preserves the semantics of `transaction` better.

> Adding `block` before `exec` fixes the blocking, so that the application stops when pressing `Ctrl-C` twice:

If the desire is to never cancel things automatically, maybe it makes sense to just skip `ROLLBACK` entirely if the query was interrupted by a `SignalException`.

For us, having to interrupt a transaction twice (or more in the case of nested transactions) is not ideal as we'd like to just `SIGTERM` a migration script and have it exit without having to worry how many times to signal it.

**Comment by thatsmydoing** [comment]
Alternatively, would it be wrong to automatically cancel a query if it's interrupted? By catching a signal error around
```
pgconn_block( 0, NULL, self ); /* wait for input (without blocking) before reading the last result */
```
for example in `async_exec` and friends? 

I saw that this was reverted for the sync API in https://github.com/ged/ruby-pg/commit/b64773f410b334e30b3732bc1d84957da725edb6

The commit also states
> async_exec allowes cancelation of queries by signals like Control+C.

Has this ever been the case? Or by cancelation here, does it mean cancelling the wait for the query but not necessarily the query itself?

In fact, the async API runs into the same problems described in the revert commit. Any signal interrupts `async_exec` but it leaves the query running so for non-transaction queries, the application code will see an error but the query continues and there's no way to retrieve the results. I'd prefer it if the query was just automatically cancelled in that case but maybe it could be an option for people who would like the existing behavior?

**Comment by larskanis** [comment]
I agree with you that it's best to cancel the running query in case of failures. It's the intention of a transaction to rollback all queries anyway and there's no use in keeping the query running any longer.

The handling of signals is not an issue on the ruby level. The issue described in commit https://github.com/ged/ruby-pg/commit/b64773f410b334e30b3732bc1d84957da725edb6 is only present on C level. The signal handling issue was (mostly) solved with the release of pg-1.1.0, which switched the default methods to the libpq async API. The `transaction` method is more or less a leftover from the state before.
